src/main/java/com/nulabinc/zxcvbn/Matching.java

```diff
@@ -12,32 +17,40 @@
         for (Map.Entry<String, String[]> frequencyListRef : Dictionary.FREQUENCY_LISTS.entrySet()) {
             String name = frequencyListRef.getKey();
             String[] ls = frequencyListRef.getValue();
-            BASE_RANKED_DICTIONARIES.put(name, buildRankedDict(ls));
+            BASE_RANKED_DICTIONARIES.put(name, buildRankedDict(Arrays.asList(ls)));
         }
     }
 
-    private final Map<String, Map<String, Integer>> rankedDictionaries;
+    protected final Map<String, Map<String, Integer>> rankedDictionaries;
 
     public Matching() {
-        this(new ArrayList<String>());
+        this(null);
     }
 
     public Matching(List<String> orderedList) {
-        if (orderedList == null) orderedList = new ArrayList<>();
-        this.rankedDictionaries = new HashMap<>();
-        for (Map.Entry<String, Map<String, Integer>> baseRankedDictionary : BASE_RANKED_DICTIONARIES.entrySet()) {
-            this.rankedDictionaries.put(
-                    baseRankedDictionary.getKey(),
-                    baseRankedDictionary.getValue());
+        this(BASE_RANKED_DICTIONARIES, orderedList);
+    }
+
+    protected Matching(Map<String, Map<String, Integer>> rankedDictionaries, List<String> orderedList) {
+        if (rankedDictionaries != null) {
+            this.rankedDictionaries = new HashMap<>(rankedDictionaries);
+        } else {
+            this.rankedDictionaries = new HashMap<>();
+        }
+        Map<String, Integer> rankedUserInputs;
+        if (orderedList != null && !orderedList.isEmpty()) {
+            rankedUserInputs = buildRankedDict(orderedList);
+        } else {
+            rankedUserInputs = Collections.emptyMap();
         }
-        this.rankedDictionaries.put("user_inputs", buildRankedDict(orderedList.toArray(new String[]{})));
+        this.rankedDictionaries.put("user_inputs", rankedUserInputs);
     }
 
     public List<Match> omnimatch(String password) {
         return new OmnibusMatcher(rankedDictionaries).execute(password);
     }
 
-    private static Map<String, Integer> buildRankedDict(String[] orderedList) {
+    protected static Map<String, Integer> buildRankedDict(List<String> orderedList) {
```

Comment:
Because we need to pass an array only when building system dictionaries. Using `Arrays.asList` does not copy underlining array, just wraps it. For orderedList (user inputs), we natively have a list. So I wanted to avoid copying it to an array. This way we save array allocation and avoid one iteration over orderedList in the hot path.